src/tokens/ERC20RebaseDistributor.sol

```diff
@@ -83,24 +86,160 @@ abstract contract ERC20RebaseDistributor is ERC20 {
+        InterpolatedValue({
+            lastTimestamp: SafeCastLib.safeCastTo32(block.timestamp),
+            lastValue: 0,
+            targetTimestamp: SafeCastLib.safeCastTo32(block.timestamp),
+            targetValue: 0
+        });
 
     /*///////////////////////////////////////////////////////////////
                             INTERNAL UTILS
     ///////////////////////////////////////////////////////////////*/
 
+    /// @notice get the current value of an interpolated value
+    function interpolatedValue(
+        InterpolatedValue memory val
+    ) internal view returns (uint256) {
+        // load state
+        uint256 lastTimestamp = uint256(val.lastTimestamp); // safe upcast
+        uint256 lastValue = uint256(val.lastValue); // safe upcast
+        uint256 targetTimestamp = uint256(val.targetTimestamp); // safe upcast
+        uint256 targetValue = uint256(val.targetValue); // safe upcast
+
+        // interpolate increase over period
+        if (block.timestamp >= targetTimestamp) {
+            // if period is passed, return target value
+            return targetValue;
+        } else {
+            // block.timestamp is within [lastTimestamp, targetTimestamp[
+            uint256 elapsed = block.timestamp - lastTimestamp;
+            uint256 delta = targetValue - lastValue;
+            return
+                lastValue +
+                (delta * elapsed) /
+                (targetTimestamp - lastTimestamp);
+        }
+    }
+
+    /// @notice called to update the number of rebasing shares.
+    /// This can happen in multiple situations:
+    /// - enterRebase()
+    /// - exitRebase()
+    /// - burn() from a rebasing address
+    /// - mint() to a rebasing address
+    /// - transfer() from a rebasing address
+    /// - transfer() to a rebasing address
+    /// - transferFrom() from a rebasing address
+    /// - transferFrom() to a rebasing address
+    /// If the number of shares is updated during the interpolation, the target share price
+    /// of the interpolation should be changed to reflect the reduced or increased number of shares
+    /// and keep a constant rebasing supply value (current & target).
+    function updateTotalRebasingShares(
+        uint256 currentRebasingSharePrice,
+        int256 sharesDelta
+    ) internal {
+        if (sharesDelta == 0) return;
+        uint256 sharesBefore = totalRebasingShares;
+        uint256 sharesAfter;
+        if (sharesDelta > 0) {
+            sharesAfter = sharesBefore + uint256(sharesDelta);
+        } else {
+            uint256 shareDecrease = uint256(-sharesDelta);
+            if (shareDecrease < sharesBefore) {
+                sharesAfter = sharesBefore - shareDecrease;
+            }
+            // else sharesAfter stays 0
+        }
+        totalRebasingShares = sharesAfter;
+
+        // reset interpolation if going to or coming from 0 rebasing supply
+        if (sharesBefore == 0 || sharesAfter == 0) {
```

Comment:
Hmmmm is it strictly necessary to check coming from and going to both? To come from 0, you first had to go to zero, except for the very first time someone enters rebase, in which case `lastValue` and `targetValue` are already set to `START_REBASING_SHARE_PRICE`; I don't believe the timestamp updates make a difference. So I think this could just be `if (sharesAfter == 0) { ...`.